tools/firrtl/generate-testbench.py

```diff
@@ -0,0 +1,129 @@
+import os
+import sys
+
+template_file = os.path.join(sys.path[0], "custom_tb_template.sv")
+
+def generate_fields(memory_cell_dicts):
+    out_str = ""
+    for memory_cell_dict in memory_cell_dicts:
+         name = memory_cell_dict['cell-name']
+         width = memory_cell_dict['WIDTH']
+         idx_size = memory_cell_dict['IDX_SIZE']
+         if (idx_size > 1):
+              out_str += f"wire [{idx_size-1}:0] {name}_addr0;\n"
+         else:
+              out_str += f"wire {name}_addr0;\n"
+         out_str += f"wire [{width-1}:0] {name}_write_data;\n"
+         out_str += f"wire {name}_write_en;\n"
+         out_str += f"wire [{width-1}:0] {name}_read_data;\n"
+         out_str += f"wire {name}_done;\n\n"
+
+    return out_str
+
+def generate_memory_dec(memory_cell_dicts):
+     out_str = ""
+     for memory_cell_dict in memory_cell_dicts:
+          name = memory_cell_dict['cell-name']
+          out_str += f'''comb_mem_d1 # (
+    .IDX_SIZE({memory_cell_dict["IDX_SIZE"]}),
+    .SIZE({memory_cell_dict["SIZE"]}),
+    .WIDTH({memory_cell_dict["WIDTH"]})
+) {name} (
+    .addr0({name}_addr0),
+    .clk(clk),
+    .done({name}_done),
+    .read_data({name}_read_data),
+    .reset(reset),
+    .write_data({name}_write_data),
+    .write_en({name}_write_en)
+);
+'''
+     return out_str
+
+def generate_main_decl(memory_cell_dicts):
+     out_str = '''main #() main (
+  .go(go),
+  .clk(clk),
+  .reset(reset),
+  .done(done)'''
+     for memory_cell_dict in memory_cell_dicts:
+          name = memory_cell_dict["cell-name"]
+          out_str += ",\n"
+          out_str += f'''  .{name}_addr0({name}.addr0),
+  .{name}_write_data({name}.write_data),
+  .{name}_write_en({name}.write_en),
+  .{name}_read_data({name}.read_data),
+  .{name}_done({name}.done)
+'''
+     out_str += "\n);"
+     return out_str
+
+def generate_readmemh(memory_cell_dicts):
+     out_str = ""
+     for memory_cell_dict in memory_cell_dicts:
+          name = memory_cell_dict["cell-name"]
+          out_str += f"  $readmemh({{DATA, \"/{name}.dat\"}}, {name}.mem);\n"
+     return out_str
+
+def generate_writememh(memory_cell_dicts):
+     out_str = ""
+     if len(memory_cell_dicts) > 0:
+          out_str += "final begin\n"
+          for memory_cell_dict in memory_cell_dicts:
+               name = memory_cell_dict["cell-name"]
+               out_str += f"    $writememh({{DATA, \"/{name}.out\"}}, {name}.mem);\n"
+          out_str += "end"
+     return out_str
+         
+def get_memory_cells(calyx_program):
+     memory_cell_dicts = []
+     with open(calyx_program) as f:
+          for line in f:
+               if ("ref" in line) and "comb_mem_d1" in line:
```

Comment:
I have to admit that this is a very hacky way to extract the memory declarations from a Calyx program, but it will do for a v1.0! At some point, it would be nice to actually parse the Calyx syntax to extract these things… maybe with yet another custom backend. But don't let that stop you from forging ahead with this string-matching version for now!

Reply:
I agree :sweat_smile: If I weren't pressed for time I definitely would've written a backend for this. I'll add it to my TODOs to fix this up after next week!